Lib/asyncio/tasks.py

```diff
@@ -829,15 +830,26 @@ def _done_callback(fut):
 
             nfuts += 1
             arg_to_fut[arg] = fut
-            fut.add_done_callback(_done_callback)
+            if fut.done():
+                # call the callback immediately instead of scheduling it
+                _done_callback(fut)
+            else:
+                all_finished = False
+                fut.add_done_callback(_done_callback)
 
         else:
             # There's a duplicate Future object in coros_or_futures.
             fut = arg_to_fut[arg]
 
         children.append(fut)
 
-    outer = _GatheringFuture(children, loop=loop)
+    if all_finished:
+        # optimization: skip creating GatheringFuture if all children completed
+        # (e.g. when all coros are able to complete eagerly)
+        outer = futures.Future(loop=loop)
+        outer.set_result([c.result for c in children])
```

Comment:
This can be removed as it will be implemented in the `_done_callback()` if everything finishes eagerly.